service/history/task/cross_cluster_task.go

```diff
@@ -208,47 +240,401 @@ func newCrossClusterTaskBase(
+	if !ok {
+		return nil, t.processingState, nil
+	}
+	ok, err := verifyTaskVersion(t.shard, t.logger, taskInfo.DomainID, signalInfo.Version, taskInfo.Version, taskInfo)
+	if err != nil || !ok {
+		return nil, t.processingState, err
+	}
+
+	return &types.CrossClusterSignalExecutionRequestAttributes{
+		TargetDomainID:    taskInfo.TargetDomainID,
+		TargetWorkflowID:  taskInfo.TargetWorkflowID,
+		TargetRunID:       taskInfo.TargetRunID,
+		RequestID:         signalInfo.SignalRequestID,
+		InitiatedEventID:  initiatedEventID,
+		ChildWorkflowOnly: taskInfo.TargetChildWorkflowOnly,
+		SignalName:        signalInfo.SignalName,
+		SignalInput:       signalInfo.Input,
+		Control:           signalInfo.Control,
+	}, t.processingState, nil
+}
+
+func (t *crossClusterSourceTask) IsValid() bool {
+	t.Lock()
+	defer t.Unlock()
+
+	return t.isValidLocked()
+}
+
+func (t *crossClusterSourceTask) isValidLocked() bool {
+	if t.processingState == processingStateInvalidated {
+		return false
+	}
+
+	domainCache := t.shard.GetDomainCache()
+	sourceEntry, err := domainCache.GetDomainByID(t.GetDomainID())
+	if err != nil {
+		// unable to tell, treat the task as valid
+		t.logger.Error("Failed to load domain entry", tag.Error(err))
+		return true
+	}
+	targetEntry, err := domainCache.GetDomainByID(t.Info.(*persistence.CrossClusterTaskInfo).TargetDomainID)
+	if err != nil {
+		return true
+	}
+
+	// pending active state is treated as valid
+	if sourceEntry.GetReplicationConfig().ActiveClusterName != t.shard.GetClusterMetadata().GetCurrentClusterName() ||
+		targetEntry.GetReplicationConfig().ActiveClusterName != t.targetCluster {
+		t.processingState = processingStateInvalidated
+		return false
+	}
+
+	return true
 }
 
-func (t *crossClusterSourceTask) GetCrossClusterRequest() *types.CrossClusterTaskRequest {
-	panic("Not implement")
+// Update records the response of processing cross cluster task at the target cluster
+// If an error is returned, the Update operation is failed, task state will remain unchanged
+// and task is still be available for polling
+// If not error is returned, operation is successful, task won't be available for polling
+// and caller should submit the task for processing.
+func (t *crossClusterSourceTask) Update(response *types.CrossClusterTaskResponse) error {
+	t.Lock()
+	defer t.Unlock()
+
+	if t.state != ctask.TaskStatePending || t.processingState != processingState(response.TaskState) {
+		// this might happen when:
+		// 1. duplicated response (shard movement in target cluster)
+		// 2. shard movement in source cluster causing task to be re-processed
+		// 3. task got invalidated during domain failover callback
+		return fmt.Errorf("unexpected task state, expected: %v, actual: %v", t.processingState, response.TaskState)
+	}
+
+	if t.GetTaskID() != response.GetTaskID() {
+		return fmt.Errorf("unexpected taskID, expected: %v, actual: %v", t.GetTaskID(), response.GetTaskID())
```

Comment:
Should it just ignore the update?

Reply:
the contract for this method is returning an error is response is not recorded. So it needs to return an error here and leave the task state unchanged. How to handle the error is up to the queue processor, and I think we just need to log the error.

For this particular case I think it should not happen otherwise it will be a bug in the queue processor impl, so logging the error is necessary I think.